engine/framework/base_jobmaster.go

```diff
@@ -226,20 +227,22 @@ func (d *DefaultBaseJobMaster) Close(ctx context.Context) error {
 
 	d.master.doClose()
 	d.worker.doClose()
+}
+
+// Close implements BaseJobMaster.Close
+func (d *DefaultBaseJobMaster) Close(ctx context.Context) error {
+	d.partialClose(ctx)
+	d.worker.cleanMessageHandler()
 	return nil
 }
 
 // NotifyExit implements BaseJobMaster interface
 func (d *DefaultBaseJobMaster) NotifyExit(ctx context.Context, errIn error) (retErr error) {
-	d.closeOnce.Do(func() {
-		err := d.impl.CloseImpl(ctx)
-		if err != nil {
-			log.Error("Failed to close JobMasterImpl", zap.Error(err))
-		}
-	})
+	d.partialClose(ctx)
```

Comment:
Could you add some comments why `partialClose` should be called twice?

Reply:
https://github.com/pingcap/tiflow/blob/dcf2404c06f05254bfd023df46dbe98b3b36c96f/engine/framework/internal/eventloop/runner.go#L68
After this PR `doGracefulExit` and `task.Close` will be excluded, so `partialClose` won't be called twice.